### PR TITLE
Make AST operators nullable

### PR DESCRIPTION
**Pull request by vyasr**
Contributes to #8980 and #8145.

This PR makes all `operator_functors` templated on the potential nullability of the inputs, allowing per-operator customization of null handling with minimal performance penalties. The default specialization in the nullable case is a null-propagating fall-through to the non-nullable case, which matches standard libcudf semantics, but any specific operator's null handling behavior can be customized by explicit specialization of the `operator_functor` template corresponding to that operator.

This PR implements `NULL_EQUAL`, `NULL_LOGICAL_AND`, and `NULL_LOGICAL_OR` operators that behave like their non-nullable counterparts when no nulls are present but can return non-null results from null inputs. The `NULL_EQUAL` operator replaces the `compare_nulls` parameter for conditional joins.

### CODE REVIEW

**Comment by vyasr** [comment]
These changes will break the JNI bindings for conditional joins, so I will update those once we're happy with the rest of the code.

**Comment by jlowe** [comment]
Happy to see these changes.  Is there any relationship to #8831?

**Comment by vyasr** [comment]
> Happy to see these changes. Is there any relationship to #8831?

Not in this PR, but those changes are the next ones on my to-do list. I started working on #8831, but the fix is slightly more involved than I initially thought so I decoupled it from this PR.

**Comment by vyasr** [comment]
@jlowe if you can double-check that [my definition of the null-tolerant EQUAL/AND/OR operators](https://github.com/rapidsai/cudf/pull/9096/files#diff-ffe693021b28f82f0147ea8f4b5730d6d569aa55c9d3109f6ead90fc02e0cdacR814) is what you were expecting that would be great. I based my operators on the [Spark documentation of null semantics](https://spark.apache.org/docs/latest/sql-ref-null-semantics.html).

File: cpp/tests/ast/transform_tests.cpp [hunk 1]
```diff
@@ -92,6 +92,22 @@ TEST_F(TransformTest, BasicAddition)
...
 
+TEST_F(TransformTest, BasicEquality)
+{
+  auto c_0   = column_wrapper<int32_t>{3, 20, 1, 50};
+  auto c_1   = column_wrapper<int32_t>{3, 7, 1, 0};
+  auto table = cudf::table_view{{c_0, c_1}};
+
+  auto col_ref_0  = cudf::ast::column_reference(0);
+  auto col_ref_1  = cudf::ast::column_reference(1);
+  auto expression = cudf::ast::operation(cudf::ast::ast_operator::EQUAL, col_ref_0, col_ref_1);
```
Comment (jlowe) [inline cpp/tests/ast/transform_tests.cpp:103]: It might be good to add a test of `EQUAL` with null inputs to validate that we didn't accidentally make `EQUAL` and `NULL_EQUAL` semantically equivalent.

**Review by jlowe — COMMENTED** [review]
> @jlowe if you can double-check that my definition of the null-tolerant EQUAL/AND/OR operators is what you were expecting that would be great.

Yep, the test cases that were added match the behavior I expected.

File: cpp/tests/ast/transform_tests.cpp (hunk 1 above)
Comment (vyasr) [reply cpp/tests/ast/transform_tests.cpp:103]: Yep can't hurt.

**Review by bdice — CHANGES_REQUESTED** [review]
Mostly minor suggestions for improvement. The design looks good overall.

One broader question: should we adopt a policy of only appending new AST operators to the end? Is there any possibility for users that rely on our C++ interface to see breakage from inserting new operators in the middle?

File: cpp/include/cudf/ast/detail/operators.hpp [hunk 2]
```diff
@@ -753,6 +780,104 @@ struct operator_functor<ast_operator::NOT> {
...
+
+  template <typename LHS,
+            typename RHS,
+            std::size_t arity_placeholder             = arity,
+            std::enable_if_t<arity_placeholder == 2>* = nullptr>
+  CUDA_DEVICE_CALLABLE auto operator()(LHS const lhs, RHS const rhs)
+    -> possibly_null_value_t<decltype(NonNullOperator{}(*lhs, *rhs)), true>
+  {
+    using Out = possibly_null_value_t<decltype(NonNullOperator{}(*lhs, *rhs)), true>;
+    return (lhs.has_value() && rhs.has_value()) ? Out{NonNullOperator{}(*lhs, *rhs)} : Out{};
```
[the remark this answers is not in the document]
Comment (vyasr) [reply cpp/include/cudf/ast/detail/operators.hpp:800]: Yes, a default constructed optional contains a null value. Thrust documentation has a lot of holes in it, I've found that for many algorithms and data structures the documentation of the C++ std equivalent is more reliable to understand expected behaviors (and have been pointed there by others in the past). In this case, the `std::optional` documentation [indicates this behavior clearly](https://en.cppreference.com/w/cpp/utility/optional), so I don't think the comment is necessary:

> The object does not contain a value in the following conditions:
> - **The object is default-initialized.**
> - The object is initialized with/assigned from a value of type std::nullopt_t or an optional object that does not contain a value.
> - The member function reset() is called.

File: cpp/tests/join/conditional_join_tests.cu [hunk 3]
```diff
@@ -31,13 +31,25 @@
 
 #include <algorithm>
 #include <random>
+#include <stdexcept>
 #include <tuple>
 #include <utility>
 #include <vector>
 
 // Defining expressions for AST evaluation is currently a bit tedious, so we
 // define some standard nodes here that can be easily reused elsewhere.
```
[the remark this answers is not in the document]
Comment (vyasr) [reply cpp/tests/join/conditional_join_tests.cu:40]: I just deleted this comment. It was more relevant when the only file-level constants were the column references required to define the `left == right` operation, but now each set of constants has more relevant comments below anyway.

File: java/src/main/java/ai/rapids/cudf/ast/BinaryOperator.java [hunk 4]
```diff
@@ -23,26 +23,29 @@
...
+  GREATER(13),            // operator >
+  LESS_EQUAL(14),         // operator <=
+  GREATER_EQUAL(15),      // operator >=
+  BITWISE_AND(16),        // operator &
+  BITWISE_OR(17),         // operator |
+  BITWISE_XOR(18),        // operator ^
+  LOGICAL_AND(19),        // operator &&
+  NULL_LOGICAL_AND(20),   // operator && but supporting Spark rules for null inputs
+  LOGICAL_OR(21),         // operator ||
+  NULL_LOGICAL_OR(22);    // operator || but supporting Spark rules for null inputs
```
[the remark this answers is not in the document]
Comment (vyasr) [reply java/src/main/java/ai/rapids/cudf/ast/BinaryOperator.java:48]: I used this wording to more closely match the [comment for `PYMOD`](https://github.com/rapidsai/cudf/pull/9096/files/4b1900c85c708d29516aad03a66fecc3b55f9033#diff-2400b4aab01582eb70dc8fdeb714c663037ea7581f748800d9709e333354024eR33). Should I change that as well?

**Comment by vyasr** [comment]
> Mostly minor suggestions for improvement. The design looks good overall.
> 
> One broader question: should we adopt a policy of only appending new AST operators to the end? Is there any possibility for users that rely on our C++ interface to see breakage from inserting new operators in the middle?

Inserting into the middle of an enum will not affect the API, but it is generally not ABI-safe, so yes this could break client code compiled against libcudf. However, at present libcudf does not even promise API compatibility between releases, so I don't think we need to worry about it. Expression evaluation is still very new and we made the last release with the expectation that we'd have to make many (potentially API- and/or ABI-breaking) changes in the subsequent release. We've made similarly ABI-breaking changes in other much more frequently used parts of the code base recently, so I think it's definitely OK to do in what's still a somewhat experimental feature in some ways.

File: java/src/main/java/ai/rapids/cudf/ast/BinaryOperator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bdice) [reply java/src/main/java/ai/rapids/cudf/ast/BinaryOperator.java:48]: I like that idea.
```cpp
PYMOD(7), // operator % using Python's sign rules for negative values
```

File: cpp/include/cudf/ast/expressions.hpp [hunk 5]
```diff
@@ -46,27 +46,35 @@ struct expression {
...
-  LOGICAL_OR,     ///< operator ||
+  ADD,               ///< operator +
+  SUB,               ///< operator -
+  MUL,               ///< operator *
+  DIV,               ///< operator / using common type of lhs and rhs
+  TRUE_DIV,          ///< operator / after promoting type to floating point
+  FLOOR_DIV,         ///< operator / after promoting to 64 bit floating point and then
+                     ///< flooring the result
+  MOD,               ///< operator %
+  PYMOD,             ///< operator % but following python's sign rules for negatives
```
Comment (bdice) [inline cpp/include/cudf/ast/expressions.hpp:57]: This operator's comment doesn't match the updates for Java bindings in 6971f52fc46205278eb2fdfa22573d51bec65e6f. It would be good to make these match. (Also capitalize the word "Python" in the Java code's comment.)
```suggestion
  PYMOD,             ///< operator % using Python's sign rules for negatives
```

File: cpp/include/cudf/ast/expressions.hpp [hunk 6]
```diff
@@ -46,27 +46,35 @@ struct expression {
...
+  BITWISE_OR,        ///< operator |
+  BITWISE_XOR,       ///< operator ^
+  LOGICAL_AND,       ///< operator &&
+  NULL_LOGICAL_AND,  ///< NULL_LOGICAL_AND(null, null) is null, NULL_LOGICAL_AND(null, true) is
+                     ///< null, NULL_LOGICAL_AND(null, false) is false, and NULL_LOGICAL_AND(valid,
+                     ///< valid) == LOGICAL_AND(valid, valid)
+  LOGICAL_OR,        ///< operator ||
+  NULL_LOGICAL_OR,   ///< NULL_LOGICAL_OR(null, null) is null, NULL_LOGICAL_OR(null, true) is true,
+                     ///< NULL_LOGICAL_OR(null, false) is null, and NULL_LOGICAL_OR(valid, valid) ==
+                     ///< LOGICAL_OR(valid, valid)
```
Comment (bdice) [inline cpp/include/cudf/ast/expressions.hpp:77]: For the `NULL_LOGICAL_AND` / `OR`, the Java comment says "using Spark rules for null inputs" and the C++ comment gives an explicit definition. It would be good to make the comments match. "Spark rules" alone is probably an insufficient description - perhaps including both pieces of information would be best? I don't have a strong opinion except that they should be consistent.

File: java/src/main/java/ai/rapids/cudf/ast/BinaryOperator.java [hunk 7]
```diff
@@ -23,26 +23,29 @@
...
-  LOGICAL_AND(18),   // operator &&
-  LOGICAL_OR(19);    // operator ||
+  ADD(0),                 // operator +
+  SUB(1),                 // operator -
+  MUL(2),                 // operator *
+  DIV(3),                 // operator / using common type of lhs and rhs
+  TRUE_DIV(4),            // operator / after promoting type to floating point
+  FLOOR_DIV(5),           // operator / after promoting to 64 bit floating point and then flooring the result
+  MOD(6),                 // operator %
+  PYMOD(7),               // operator % using python's sign rules for negatives
```
Comment (bdice) [inline java/src/main/java/ai/rapids/cudf/ast/BinaryOperator.java:33]:
```suggestion
  PYMOD(7),               // operator % using Python's sign rules for negatives
```

**Review by bdice — CHANGES_REQUESTED** [review]
Minor changes requested. Otherwise looks good to me.

File: cpp/include/cudf/ast/expressions.hpp (hunk 6 above)
Comment (vyasr) [reply cpp/include/cudf/ast/expressions.hpp:77]: Discussed offline: I don't want them to match exactly because the C++ code will be read by libcudf devs who aren't necessarily familiar with Spark rules, unlike our Java developers. However, I've updated these docstrings to explicitly indicate that the behavior described here corresponds to Spark rules, which should help bridge the gap.

**Review by bdice — APPROVED** [review]

**Review by nvdbaranec — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge
